packages/@aws-cdk/aws-apigateway/lib/gateway-response.ts

```diff
@@ -0,0 +1,210 @@
+import { Construct, IResource, Resource } from '@aws-cdk/core';
+import { CfnGatewayResponse } from './apigateway.generated';
+import { IRestApi } from './restapi';
+
+/**
+ * Represents gateway response resource.
+ */
+export interface IGatewayResponse extends IResource {
+}
+
+/**
+ * Properties for a new gateway response.
+ */
+export interface GatewayResponseProps extends GatewayResponseOptions {
+  /**
+   * Rest api resource to target.
+   */
+  readonly restApi: IRestApi;
+}
+
+/**
+ * Options to add gateway response.
+ */
+export interface GatewayResponseOptions {
+  /**
+   * Response type to associate with gateway response.
+   * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/supported-gateway-response-types.html
+   */
+  readonly type: ResponseType;
+
+  /**
+   * Http status code for response.
+   * @default - standard http status code for the response type.
+   */
+  readonly statusCode?: string;
+
+  /**
+   * Custom headers parameters for response.
+   * @default - no headers
+   */
+  readonly responseHeaders?: { [key: string]: string };
+
+  /**
+   * Custom templates to get mapped as response.
+   * @default - Response from api will be returned without applying any transformation.
+   */
+  readonly templates?: { [key: string]: string };
+
+}
+
+/**
+ * Configure the response received by clients, produced from the API Gateway backend.
+ *
+ * @resource AWS::ApiGateway::GatewayResponse
+ */
+export class GatewayResponse extends Resource implements IGatewayResponse {
+  constructor(scope: Construct, id: string, props: GatewayResponseProps) {
+    super(scope, id);
+
+    const resource = new CfnGatewayResponse(this, 'Resource', {
+      restApiId: props.restApi.restApiId,
+      responseType: props.type.responseType,
+      responseParameters: this.buildResponseParameters(props.responseHeaders),
+      responseTemplates: props.templates,
+      statusCode: props.statusCode,
+    });
+
+    this.node.defaultChild = resource;
+  }
+
+  private buildResponseParameters(responseHeaders?: { [key: string]: string }): { [key: string]: string } | undefined {
+    if (!responseHeaders) {
+      return undefined;
+    }
+
+    const responseParameters: { [key: string]: string } = {};
+    for (const [header, value] of Object.entries(responseHeaders)) {
+      responseParameters[`gatewayresponse.header.${header}`] = value;
+    }
+    return responseParameters;
+  }
+}
+
+/**
+ * Supported types of gateway responses.
+ * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/supported-gateway-response-types.html
+ */
+export class ResponseType {
```

Comment:
@nirvana124 Is this class automatically generated or written? How is this generated if automatically generated?